### ISSUE TITLE
ConvTranspose def

### ISSUE THREAD

**Issue by jerryzh168** [onnx/onnx#102]
I feel attributes of ConvTranspose is not really well defined. Specifically why do we have "output_shape" as an argument, I think it should be something we infer from the size of input right? Did we discuss about it before?
https://github.com/onnx/onnx/blob/master/onnx/defs/nn/defs.cc#L145

**Comment by fmassa** [comment]
The reason for `output_shape` is that when you perform a convolution with stride > 1, you might lose boundary pixels, so a 5x5 image with a stride os 2 will be of size 2x2. So there is no way of knowing only from kernel sizes / strides and the original input size (2x2 now) that you want your `ConvTranspose` to have a size of 5x5. This issue gives some more information on that https://github.com/torch/nn/pull/552

**Comment by jerryzh168** [comment]
@fmassa Yeah I'm not suggesting to remove `output_shape`, I just feel `adj` is more consistent with the other attributes in the operator.

**Comment by ebarsoum** [comment]
I don't understand, what adj does?

**Comment by jerryzh168** [comment]
`a` in the following:
`o = s (i - 1) + a + k - 2p.` where `o` is output size, `s` is stride, `i` is input size, `a` is adj, `k` is kernel and `p` is pad.

**Comment by jerryzh168** [comment]
Please refer to http://deeplearning.net/software/theano/tutorial/conv_arithmetic.html#transposed-convolution-arithmetic, 

A transposed convolution specified by
input size i,
kernel size k,
stride s,
padding size p,
has an output size given by
o = s (i - 1) + a + k - 2p, a in {0, ..., s - 1}
where a is a user-specified quantity used to distinguish between the s different possible output sizes.

**Comment by ebarsoum** [comment]
I know transpose convolution and how to compute the output size, I don't understand what `adj` mean and why it is better than output_shape?

**Comment by ezyang** [comment]
Are adj and output_size equivalently expressive, or is one more expressive than the other?

**Comment by jerryzh168** [comment]
@ebarsoum `adj` is the `a` in the `output_shape` equation above. Because `output_shape` looks like a value derived from these settings. @ezyang They are equally expressive.

**Comment by dzhulgakov** [comment]
Reviving this thread. After checking existing implementations, it seems that people do different stuff and do support `adj` fairly often instead of `output_shape`:
Caffe2: adj only https://github.com/caffe2/caffe2/blob/master/caffe2/operators/conv_transpose_unpool_op_base.h#L45
PyTorch: adj only  https://github.com/pytorch/pytorch/blob/25b166ed1f58fe1ba255cf218f8b44bf7cc9bb5b/torch/nn/modules/conv.py#L473
MXNet: both
https://mxnet.incubator.apache.org/api/python/symbol.html#mxnet.symbol.Deconvolution
TensorFlow: output_shape

I'm a bit biased towards `adj` because it makes things work even with variable input size (though probably nobody is going to invoke it like that).

Options:
1) replace `output_shape` with `adj`
2) allow both (like mxnet) but specify that only one needs to be specified

**Comment by ezyang** [comment]
I recently was working on PyTorch's CuDNN bindings and I also noticed that you often want to support both APIs. But what I found, like as @dzhulgakov says, is that adj is most widely used for transposed convolution, and output shape is most widely used for backwards convolution.

There is one downside to adj, which is that it is more straightforward to compute output shape from adj than vice versa, so doing the reverse transformation is a bit annoying to do. 

BTW, adj is called `output_padding` in PyTorch.
